src/Connection/DefaultConnectionPool.php

```diff
@@ -36,73 +36,106 @@ public function __construct(?Connector $connector = null)
-                    throw new TimeoutException(\sprintf("TLS handshake with '%s' @ '%s' timed out, took longer than " . $request->getTlsHandshakeTimeout() . ' ms', $authority, $socket->getRemoteAddress()->toString())); // don't pass $e
+                    throw new TimeoutException(\sprintf("Connection to '%s' timed out, took longer than " . $request->getTcpConnectTimeout() . ' ms', $authority)); // don't pass $e
                 }
-            }
 
-            if (!isset($this->connections[$authority])) {
-                $this->connections[$authority] = [];
-            }
+                if ($isHttps) {
+                    try {
+                        $tlsState = $socket->getTlsState();
+                        if ($tlsState === EncryptableSocket::TLS_STATE_DISABLED) {
+                            $tlsCancellationToken = new CombinedCancellationToken($cancellation, new TimeoutCancellationToken($request->getTlsHandshakeTimeout()));
+                            yield $socket->setupTls($tlsCancellationToken);
+                        } elseif ($tlsState !== EncryptableSocket::TLS_STATE_ENABLED) {
+                            throw new SocketException('Failed to setup TLS connection, connection was in an unexpected TLS state (' . $tlsState . ')');
+                        }
+                    } catch (StreamException $exception) {
+                        throw new SocketException(\sprintf("Connection to '%s' closed during TLS handshake", $authority), 0, $exception);
+                    } catch (CancelledException $e) {
+                        // In case of a user cancellation request, throw the expected exception
+                        $cancellation->throwIfRequested();
+
+                        // Otherwise we ran into a timeout of our TimeoutCancellationToken
+                        throw new TimeoutException(\sprintf("TLS handshake with '%s' @ '%s' timed out, took longer than " . $request->getTlsHandshakeTimeout() . ' ms', $authority, $socket->getRemoteAddress()->toString())); // don't pass $e
+                    }
+                }
+
+                $connection = new Http1Connection($socket);
+
+                $connections = &$this->connections;
+                $connection->onClose(static function () use (&$connections, $key, $index) {
+                    unset($connections[$key][$index]);
+
+                    if (empty($connections[$key])) {
+                        unset($connections[$key]);
+                    }
+                });
+
+                return $connection;
+            });
+
+            $promise->onResolve(function (?\Throwable $exception, ?Connection $connection) use ($key, $index): void {
+                if ($exception) {
+                    unset($this->connections[$key][$index]);
+
+                    if (empty($this->connections[$key])) {
+                        unset($this->connections[$key]);
+                    }
+                    return;
+                }
```

Comment:
We only push the connection to the array two lines below, so there shouldn't be any entry here to delete.

Reply:
The promise is pushed to the array when it's created, so that needs to be deleted.